File: 705 - design hashset.cpp
class MyHashSet {
public:
    vector<vector<int>> storage;
    int buckets;
    int bucketItems;
    /** Initialize your data structure here. */
    MyHashSet() {
        buckets = 1000;
        bucketItems = 1000;
        storage.resize(buckets,vector<int>());
        
    }
    
    int getBucket(int key){
        return key % buckets;
    }
    
    int getBucketItem(int key){
        return key / bucketItems;
    }
    
    void add(int key) {
        int bucket = getBucket(key);
        int bucketItem = getBucketItem(key);
        if(storage[bucket].empty()){
            if(bucket == 0){
                storage[bucket].resize(bucketItems+1,0);
            }
            else{
                storage[bucket].resize(bucketItems,0);
            }
        }
        storage[bucket][bucketItem] = 1;
    }
    
    void remove(int key) {
        int bucket = getBucket(key);
        int bucketItem = getBucketItem(key);
        if(!storage[bucket].empty()){
            storage[bucket][bucketItem] = 0;
        }
        
    }
    
    /** Returns true if this set contains the specified element */
    bool contains(int key) {
        int bucket = getBucket(key);
        int bucketItem = getBucketItem(key);
        if(storage[bucket].empty()){
            return false;
        }
        return storage[bucket][bucketItem];
    }
};

/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet* obj = new MyHashSet();
 * obj->add(key);
 * obj->remove(key);
 * bool param_3 = obj->contains(key);
 */

File: 705_Design_HashSet.py
'''
## Problem 1:(https://leetcode.com/problems/design-hashset/)

## Author: Neha Doiphode
## Date  : 06-21-2022

## Title:
    Design a HashSet without using any built-in hash table libraries.

## Description:
    Implement MyHashSet class:
    void add(key) Inserts the value key into the HashSet.
    bool contains(key) Returns whether the value key exists in the HashSet or not.
    void remove(key) Removes the value key in the HashSet. If key does not exist in the HashSet, do nothing.



## Example 1:
    Input
        ["MyHashSet", "add", "add", "contains", "contains", "add", "contains", "remove", "contains"]
        [[], [1], [2], [1], [3], [2], [2], [2], [2]]

    Output
        [null, null, null, true, false, null, true, null, false]

    Explanation
        MyHashSet myHashSet = new MyHashSet();
        myHashSet.add(1);      // set = [1]
        myHashSet.add(2);      // set = [1, 2]
        myHashSet.contains(1); // return True
        myHashSet.contains(3); // return False, (not found)
        myHashSet.add(2);      // set = [1, 2]
        myHashSet.contains(2); // return True
        myHashSet.remove(2);   // set = [1]
        myHashSet.contains(2); // return False, (already removed)

## Constraints:
        0 <= key <= 106
        At most 104 calls will be made to add, remove, and contains.

## Time complexity:
    O(n/k), where n is number of possible values to be added in the hashset.
                  k is number of pre-defined buckets.

            In worst case we might have to traverse, all buckets and all elements present in the bucket.

## Space complexity:
   O(n + k), where n is number pre-defined buckets
                   k is number of elements to be added in the hashset.

'''


class MyHashSet:
    def __init__(self):
        self._buckets = 1000
        self._bucket_items = 1000
        self._storage = [None] * self._buckets

    def _get_bucket(self, key) -> int:
        return int(key % self._buckets)

    def _get_bucket_item(self, key) -> int:
        return int(key / self._bucket_items)

    def add(self, key: int) -> None:
        bucket = self._get_bucket(key)
        bucket_item = self._get_bucket_item(key)

        if self._storage[bucket] == None:
            if bucket == 0:
                self._storage[bucket] = [False] * (self._bucket_items + 1)
            else:
                self._storage[bucket] = [False] * self._bucket_items

        self._storage[bucket][bucket_item] = True

    def remove(self, key: int) -> None:
        bucket = self._get_bucket(key)
        bucket_item = self._get_bucket_item(key)

        if self._storage[bucket] == None:
            return

        self._storage[bucket][bucket_item] = False

    def contains(self, key: int) -> bool:
        bucket = self._get_bucket(key)
        bucket_item = self._get_bucket_item(key)
        answer = False if self._storage[bucket] == None else self._storage[bucket][bucket_item]
        return answer


# Driver code
myHashSet = MyHashSet()
myHashSet.add(1)      # set = [1]
myHashSet.add(2)      # set = [1, 2]
print(myHashSet.contains(1)) # return True
print(myHashSet.contains(3)) # return False, (not found)
myHashSet.add(2)      # set = [1, 2]
print(myHashSet.contains(2)) # return True
myHashSet.remove(2)   # set = [1]
print(myHashSet.contains(2)) # return False, (already removed)
